tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -1454,38 +1454,164 @@ Status Converter::RenameAndMarkOutputTensors(
 
-    // making sure that the requested TRT algorithm ID doesn't go above the
-    // max value accepted.
-    selection[0] = std::min(algorithm_id_, nbChoices);
-    return 1;
+    absl::optional<int64_t> forced_algorithm_id  = absl::nullopt;
+
+    if (! allowed_algorithm_ids.empty()) {
+      assert(allowed_algorithm_ids.size() == 1);
+      forced_algorithm_id = *allowed_algorithm_ids.begin(0);
+      // Making sure that the requested TRT algorithm ID doesn't go above
+      // the max value accepted.
+      forced_algorithm_id = std::min(forced_algorithm_id.value(),
+                                     static_cast<int64_t>(nbChoices));
+
+      VLOG(1) << "Forcing TRT algorithm selection to: ID = "
+              << forced_algorithm_id.value();
+      selection[nbSelections++] = forced_algorithm_id.value();
+    }
+
+    auto forced_algorithm_id_is_banned = [](int64_t _algorithm_id){
+      VLOG(1) << "Impossible to force TensorRT tactic selection. All "
+              << "attempts have been rejected. Please select a different "
+              << "value for `TF_TRT_FIXED_ALGORITHM_ID`. Received: "
+              << _algorithm_id << ".";
+      assert(false);
+    };
+
+    // Disable all disallowed tactics
+    for (auto i = 0; i < nbChoices; i++) {
+      if (algoChoices[i] == nullptr) {
+        continue;
+      }
+
+      int64_t tacticId = algoChoices[i]->getAlgorithmVariant().getTactic();
+
+      if (disallowed_algorithm_ids.find(tacticId) !=
+          disallowed_algorithm_ids.end()) {
+        if (!forced_algorithm_id.has_value()) {
+          VLOG(1) << "Rejecting a disallowed tactic: " << tacticId
+                  << " for node " << algoContext.getName();
+          continue;
+        } else if (forced_algorithm_id.value() != i) {
+          continue;
+        }
+        forced_algorithm_id_is_banned(i);
+      }
+
+#if !IS_TRT_VERSION_GE(8, 0, 0, 0)
+      int64_t implementation =
+          algoChoices[i]->getAlgorithmVariant().getImplementation();
+      nvinfer1::TensorFormat format =
+          algoChoices[i]->getAlgorithmIOInfo(0).getTensorFormat();
+      assert(tacticId != static_cast<int64_t>(Tactic::kINVALID_TACTIC));
+
+      // Reject shuffle node when input format is 32-wide channel vectorized
+      // row major FP32 format
+      if (implementation == static_cast<int64_t>(LayerImpl::kSHUFFLE) &&
+          format == nvinfer1::TensorFormat::kCHW32) {
+        if (!forced_algorithm_id.has_value()) {
+          VLOG(1) << "Rejecting a disallowed tactic: " << tacticId
+                  << " for node " << algoContext.getName()
+                  << " with implementation " << implementation
+                  << " and input format " << static_cast<int32_t>(format);
+          continue;
+        } else if (forced_algorithm_id.value() != i) {
+          continue;
+        }
+        forced_algorithm_id_is_banned(i);
+      }
+#endif  // !IS_TRT_VERSION_GE(8, 0, 0, 0)
+      if (!forced_algorithm_id.has_value()) {
```

Comment:
remove this if also, only leaving line 1603

Reply:
This proposal doesn't work.
It would be equal to insert multiple times : `selection[nbSelections++] = forced_algorithm_id.value();` in case it's defined.

Reply:
If you remove line 1548 that assigns the tactic to selection, you should be able to remove the if statement here, without inserting the id passed in through the environment variable multiple times.

Reply:
Fixed